### PR TITLE
CORDA-2876: Integrate the DJVM with the Corda Node.

### PR DESCRIPTION
**Pull request by chrisr3**
Teach Corda to perform contract verification inside the DJVM's sandbox.

The breaks in API Stability are because `LedgerTransaction` is no longer annotated with `@CordaSerializable`. However, this was originally done only for the sake of the external verifier:
```
// TODO LedgerTransaction is not supposed to be serialisable as it references attachments, etc. The verification logic
// currently sends this across to out-of-process verifiers. We'll need to change that first.
```

No DJVM classes are leaking into the `api-current.txt` file.

I have configured the Corda node _not_ to enable the DJVM unless the system property `net.corda.djvm=true`, and also not to override `co.paralleluniverse.fibers.verifyInstrumentation` in devMode if this system property has been set explicitly.

DemoBench now launches nodes with
```java
-Dnet.corda.djvm=true
-Dco.paralleluniverse.fibers.verifyInstrumentation=false
```

### CODE REVIEW

File: constants.properties [hunk 1]
```diff
@@ -30,6 +29,7 @@ snakeYamlVersion=1.19
 caffeineVersion=2.7.0
 metricsVersion=4.1.0
 metricsNewRelicVersion=1.1.1
+djvmVersion=1.0-SNAPSHOT
```
Comment (IgorNittoR3) [inline constants.properties]: Does this point to a persistent SNAPSHOT version?

Comment (chrisr3) [reply constants.properties]: Not yet. But I'm also still pushing new DJVM SNAPSHOTs to the Artifactory manually.so this shouldn't be a big concern.

File: build.gradle [hunk 2]
```diff
@@ -335,6 +336,7 @@ allprojects {
         maven { url "$artifactory_contextUrl/corda-dependencies" }
         maven { url 'https://jitpack.io' }
         maven { url 'https://repo.gradle.org/gradle/libs-releases' }
+        maven { url "$artifactory_contextUrl/corda-dev" }
```
Comment (blsemo) [inline build.gradle:339]: The DJVM should probably be released into `corda-dependencies`.

File: constants.properties (hunk 1 above)
Comment (blsemo) [reply constants.properties]: Hm. Depending on a snapshot version in Corda is a bit yucky.

Comment (chrisr3) [reply constants.properties]: Upgraded to 1.0-RC01.

File: core-tests/src/test/kotlin/net/corda/coretests/crypto/PartialMerkleTreeTest.kt [hunk 3]
```diff
@@ -307,37 +306,37 @@ class PartialMerkleTreeTest {
 
         val pmt = PartialMerkleTree.build(merkleTree, listOf<SecureHash>(SecureHash.sha256("1"), SecureHash.sha256("5"), SecureHash.sha256("0"), SecureHash.sha256("19")))
         // First leaf.
-        assertEquals(0, pmt.accessLeafIndex(SecureHash.sha256("0")))
+        assertEquals(0, pmt.leafIndex(SecureHash.sha256("0")))
```
Comment (IgorNittoR3) [inline core-tests/src/test/kotlin/net/corda/coretests/crypto/PartialMerkleTreeTest.kt:309]: Any particular reason for changing  `PartialMerkleTree` ?

Comment (chrisr3) [reply core-tests/src/test/kotlin/net/corda/coretests/crypto/PartialMerkleTreeTest.kt:309]: Yes. I needed to invoke `leafIndex()` from outside the `core` module.
```diff
+    @CordaInternal
     @Throws(MerkleTreeException::class)
-    internal fun leafIndex(leaf: SecureHash): Int {
+    fun leafIndex(leaf: SecureHash): Int {
```

File: build.gradle (hunk 2 above)
Comment (chrisr3) [reply build.gradle:339]: And it _will_ be, Grasshopper! :owl: But it's currently only using SNAPSHOT and RC builds.

File: node/capsule/build.gradle [hunk 4]
```diff
@@ -34,28 +36,64 @@ capsule {
...
         "$rootDir/config/dev/log4j2.xml",
         'NOTICE' // Copy CDDL notice
     )
     from configurations.capsuleRuntime.files.collect { zipTree(it) }
     with jar
 
+    // The DJVM will share most of its dependencies with the node, but any extra ones that it needs
+    // are listed in the node's "deterministic" configuration and copied into a djvm subdirectory.
+    //
+    // Gradle may not resolve exactly rhe same transitive dependencies for both the runtimeClasspath
```
Comment (blsemo) [inline node/capsule/build.gradle]: Typo - I presume you meant "the"

File: node/djvm/src/main/kotlin/net/corda/node/djvm/CommandBuilder.kt [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+class CommandBuilder : Function<Array<Any?>, List<CommandWithParties<CommandData>>> {
+    @Suppress("unchecked_cast")
+    override fun apply(inputs: Array<Any?>): List<CommandWithParties<CommandData>> {
+        val signers = inputs[0] as? List<List<PublicKey>> ?: emptyList()
+        val commandsData = inputs[1] as? List<CommandData> ?: emptyList()
+        val partialMerkleLeafIndices = inputs[2] as? IntArray
+
+        return if (partialMerkleLeafIndices != null) {
+            check(commandsData.size <= signers.size) {
```
Comment (blsemo) [inline node/djvm/src/main/kotlin/net/corda/node/djvm/CommandBuilder.kt:20]: Is this always a valid assumption? Does every command have to align to one signer?

File: node/src/main/kotlin/net/corda/node/services/config/ConfigUtilities.kt [hunk 6]
```diff
@@ -59,8 +57,11 @@ object ConfigHelper {
                 .withFallback(defaultConfig)
                 .resolve()
 
-        val entrySet = finalConfig.entrySet().filter { entry -> entry.key.contains("\"") }
-        for ((key) in entrySet) {
+        val entrySet = finalConfig.entrySet().filter { entry ->
+            // System properties can reasonably be expected to contain '.'.
+            with(entry.key) { contains("\"") && !matches("^systemProperties\\.\"[^\"]++\"\$".toRegex()) }
```
Comment (blsemo) [inline node/src/main/kotlin/net/corda/node/services/config/ConfigUtilities.kt:61]: Can you add a comment what this new RegEx is trying to achieve, please?

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockServices.kt [hunk 7]
```diff
@@ -490,4 +490,4 @@ fun <T : SerializeAsToken> createMockCordaService(serviceHub: MockServices, serv
         }
     }
     return MockAppServiceHubImpl(serviceHub, serviceConstructor).serviceInstance
-}
\ No newline at end of file
+}
```
Comment (blsemo) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/MockServices.kt:493]: Really? Adding a file to add a new line at the end?

**Review by blsemo — COMMENTED** [review]
I have given up on reading all the serializers in detail at some point.

File: node/capsule/build.gradle (hunk 4 above)
Comment (chrisr3) [reply node/capsule/build.gradle]: :+1:

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockServices.kt (hunk 7 above)
Comment (chrisr3) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/MockServices.kt:493]: I blame IntelliJ...

File: node/src/main/kotlin/net/corda/node/services/config/ConfigUtilities.kt (hunk 6 above)
Comment (chrisr3) [reply node/src/main/kotlin/net/corda/node/services/config/ConfigUtilities.kt:61]: It's either _exactly_ matching entries of the form
```java
systemProperties."nonempty-text-without-any-quotes-in"
```
or it's trying to summon Gozer the Gozerian to bring about the End of the World.

File: node/djvm/src/main/kotlin/net/corda/node/djvm/CommandBuilder.kt (hunk 5 above)
Comment (chrisr3) [reply node/djvm/src/main/kotlin/net/corda/node/djvm/CommandBuilder.kt:20]: Logic copied(?) from `net.corda.core.internal.deserialiseCommands()`.

**Review by mnesbit — APPROVED** [review]
OK

**Pull request merged**